File: lim/이진탐색/Parametric Search - 이차원 배열의 오름차순 정리.py
'''
행렬에서 한 행을 고정시키면 각 행은 i의 배수이므로 순서를 구하는 특정수에 i를 나눠 특정수보다 작은 수 개수를 셀수있다
그외 자잘한 조건이 많다
'''

import math
import sys


def sol(n, k):
    def get_order(num):
        order = 1
        for i in range(1, n+1):
            lt_count = math.ceil(min(num, n*i+0.1) / i) - 1
            order += lt_count
        return order

    def search(target_order):
        left, right = 0, n**2
        answer = -1
        while left <= right:
            mid = (left + right) // 2
            min_order = get_order(mid)
            max_order = get_order(mid+1) - 1
            # print('order', min_order, max_order, 'left', left, 'right', right, 'mid', mid)
            if min_order <= target_order and target_order <= max_order:
                answer = mid
                break
            elif target_order < min_order:
                right = mid - 1
            else:
                left = mid + 1
        return answer

    print(search(k))


n = int(sys.stdin.readline())
k = int(sys.stdin.readline())

sol(n, k)

File: lim/이진탐색/Parametric_Search_이차원_배열의_오름차순_정리.java
package lim.이진탐색;

/*
 * 정답 숫자 num을 가정한다
 * num보다 작은 숫자들 개수를 구한다
 * "보다 작은 숫자 개수 + 1 <= 주어진 순서"인 최대 num을 구하면 정답
 */
import java.util.*;
import java.io.*;

public class Parametric_Search_이차원_배열의_오름차순_정리 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int n, targetOrder;

    static long getUnderCount(int rIdx, long num) {
        int left = 0, right = n - 1;
        int idx = 0;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (((long) mid + 1) * (rIdx + 1) < num) {
                left = mid + 1;
                idx = mid;
            } else {
                right = mid - 1;
            }
        }
        return idx + 1;
    }

    static long getUnderCount(long num) {
        long count = 0;
        for (int i = 0; i < n; i++) {
            count += getUnderCount(i, num);
        }
        return count;
    }

    static boolean check(long num) {
        long count = getUnderCount(num);
        return count + 1 <= targetOrder;
    }

    static void sol() throws IOException {
        long left = 1, right = 10_000_000_000L;
        long answer = 0;
        while (left <= right) {
            long mid = (left + right) / 2;
            if (check(mid)) {
                left = mid + 1;
                answer = mid;
            } else {
                right = mid - 1;
            }
        }
        bw.write(answer + "");
    }

    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        targetOrder = Integer.parseInt(st.nextToken());
        sol();
        br.close();
        bw.flush();
        bw.close();
    }
}
